### PR TITLE
CNOT doesn't decompose to CNOT

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
Up to and including Pennylane 0.34, the CNOT gate raised a `DecompositionUndefinedError` in its `decomposition` method. We added some smart handling of decompositions for controlled operators that accidentally started decomposing CNOT into itself. 

**Description of the Change:**
Make it so `_decompose_pauli_x_based_no_control_values` no longer provides a decomposition for `CNOT` (it currently just sees a controlled op with a base of `PauliX`, and decomposes to `CNOT`). The decomposition functions then follow another pathway and return another decomposition for CNOT.

**Benefits:**
This fringe but odd behaviour doesn't happen anymore

**Possible Drawbacks:**
I tested it by adding a check to `assert_valid`, maybe this problem is too much of an edge case to be in `assert_valid`. But with the controlled ops finding their own decompositions instead of having them hardcoded, I guess it could happen again 🤷‍♀️ 

**Related GitHub Issues:**
#5711

### CODE REVIEW

**Comment by lillian542** [comment]
[sc-63619]

File: tests/ops/op_math/test_controlled.py [hunk 1]
```diff
@@ -1930,13 +1930,14 @@ def make_ops():
...
             *qml.CRot(0.111, 0.222, 0.333, wires=[1, 2]).decomposition(),
-            qml.CNOT(wires=[1, 2]),
+            *qml.CNOT(wires=[1, 2]).decomposition(),
             *qml.CY(wires=[1, 4]).decomposition(),
             *qml.CZ(wires=[1, 0]).decomposition(),
             qml.PauliX(wires=1),
         ]
         assert len(tape) == 9
         expanded = tape.expand(stop_at=lambda obj: not isinstance(obj, Controlled))
+        print(expanded.circuit)
```
Comment (dwierichs) [inline tests/ops/op_math/test_controlled.py:1940]:
```suggestion
```

File: tests/ops/op_math/test_controlled.py [hunk 2]
```diff
@@ -2183,8 +2184,8 @@ def ansatz(weights, wires):
 
         tape = QuantumScript.from_queue(q_tape)
         tape = tape.expand(depth=1, stop_at=lambda obj: not isinstance(obj, Controlled))
-        assert len(tape.operations) == 10
-        assert all(o.name in {"CNOT", "CRX", "Toffoli"} for o in tape.operations)
+        assert len(tape.operations) == 12
+        assert all(o.name in {"ControlledPhaseShift", "CRX", "Toffoli"} for o in tape.operations)
```
Comment (dwierichs) [inline tests/ops/op_math/test_controlled.py:2188]: Just to make sure I understand: The requirement for a decomposition is not to imply any ordering of "complex gates" to "simpler gates", but just should allow us to keep decomposing until external stopping criteria (like those of a device) trigger? :)

File: pennylane/ops/functions/assert_valid.py [hunk 3]
```diff
@@ -56,6 +56,9 @@ def _check_decomposition(op):
         expand = op.expand()
 
         assert isinstance(decomp, list), "decomposition must be a list"
+        assert op.__class__ not in [
+            decomp_op.__class__ for decomp_op in decomp
+        ], "an operator should not be included in its own decomposition"
```
Comment (dwierichs) [inline pennylane/ops/functions/assert_valid.py:61]: I actually would have expected some other operation popping up to have this bug here :D

Comment (dwierichs) [inline pennylane/ops/functions/assert_valid.py:61]: Could do
```suggestion
        assert op.__class__ not in {
            decomp_op.__class__ for decomp_op in decomp
        }, "an operator should not be included in its own decomposition"
```

**Review by dwierichs — APPROVED** [review]
Thanks @lillian542, this looks good to me! Good catch :) 

I just had a tiny suggestion, would ask to remove a `print` line, and wanted to confirm I understand the background correctly with a question :)

**edit:** Actually, this change might warrant a breaking change entry in the changelog, given the "failing" tape expansion tests and that `CNOT`s are omnipresent in standard circuits.

File: tests/ops/op_math/test_controlled.py (hunk 2 above)
Comment (lillian542) [reply tests/ops/op_math/test_controlled.py:2188]: Exactly!

**Comment by lillian542** [comment]
> edit: Actually, this change might warrant a breaking change entry in the changelog, given the "failing" tape expansion tests and that CNOTs are omnipresent in standard circuits.

Agreed, I want to bring this up in stand up tomorrow and make sure it's something we actually want (I think it is, just want to confirm). Hopefully most people are decomposing in ways that include `CNOT` as a supported op and almost no one will notice. 

But it could be between this and `max_expansion=10` in the old device API that someone somewhere is decomposing CNOT into CNOT 10 times and then proceeding 😅 And that person will be surprised when all the CNOTs disappear from their circuits.

**Comment by trbromley** [comment]
So just to summarise our discussion in the standup, here we are effectively reverting to behaviour in Q1?

**Comment by trbromley** [comment]
Is this to be reopened later on?

**Comment by lillian542** [comment]
Oops! I forgot to change the target branch, and this happens when the target branch is deleted. It also happened for one of Matthew's PRs last time I was managing the release, and I think he had to re-create it, we weren't able to re-open it. But we can open a new PR from the same branch.

**Comment by lillian542** [comment]
Now re-opened as #6039

**Pull request closed without merging**